Sources/Wrappers/AngouriMath.FSharp/Matrices.fs

```diff
@@ -3,45 +3,101 @@
 open AngouriMath
 open AngouriMath.FSharp.Core
 open AngouriMath.FSharp.Functions
+open System.Linq
 
-let (|*) a b = (parsed a * parsed b).InnerSimplified |> asMatrix
+/// Represents the length type of a matrix
+type Lengths =
+    | Any
+    | Invalid
+    | Fixed of int
 
-let (|**) a b = (parsed a ** parsed b).InnerSimplified |> asMatrix
+/// Creates a matrix from a list of lists
+/// of objects (which are parsed into Entities)
+let matrix x = 
+    let rec columnCount (x : 'T list list) =
+        match x with
+        | [] -> Any
+        | hd::tl ->
+            match columnCount tl with
+            | Any -> Fixed(hd.Length)
+            | Invalid -> Invalid
+            | Fixed len -> if len = hd.Length then Fixed(len) else Invalid
+            
+    let parseListOfLists li =
+        [ for row in li do yield [ for el in row do yield (parsed el) ] ]
 
-let (|+) a b = (parsed a + parsed b).InnerSimplified |> asMatrix
+    match columnCount x with
+    | Any | Invalid -> raise ParseException
+    | Fixed _ -> MathS.Matrix(array2D (parseListOfLists x))
 
-let (|-) a b = (parsed a - parsed b).InnerSimplified |> asMatrix
-
-let (|/) a b = (parsed a / parsed b).InnerSimplified |> asMatrix
+/// Creates a column vector from a 1-dimensional list
+let vector li =
+    MathS.Vector([ for el in li do yield parsed el ].ToArray())
 
+/// If the provided entity is a matrix,
+/// it is returned downcasted. Otherwise,
+/// a 1x1 matrix containing the entity is returned.
+let asMatrix a = 
+    match parsed a with
+    | :? Entity.Matrix as m -> m
+    | other -> vector [other]
 
 /// Creates a square 2x2 matrix, with four elements
 /// in the following order: left-top, right-top,
 /// left-bottom, right-bottom
 let matrix2x2 leftTop rightTop leftBottom rightBottom = 
     matrix [[leftTop; rightTop]; [leftBottom; rightBottom]]
-
+ 
 /// Creates a square 3x3 matrix, going first from left
 /// to right, then from top to bottom (the same way
 /// people in Europe read)
 let matrix3x3 leftTop middleTop rightTop leftMiddle middleMiddle rightMiddle leftBottom middleBottom rightBottom =
     matrix [[leftTop; middleTop; rightTop]; [leftMiddle; middleMiddle; rightMiddle]; [leftBottom; middleBottom; rightBottom]]
-
+ 
 /// Creates a vector of two elements
 let vector2 a b = vector [a; b]
-
+ 
 /// Creates a vector of three elements
 let vector3 a b c = vector [a; b; c]
 
+/// Returns a multiline string representation of matrix
+let printedMatrix (x: AngouriMath.Entity.Matrix) = x.ToString(true)
```

Comment:
I would name this `matrixToString`.

Reply:
But it would create confusion with normal `ToString()` then

Reply:
`matrixToMultilineString`?

Reply:
Hmmmm... let's keep it as it is for now, until it's clear what a good name is